File: android/app/src/main/java/com/emwaver/emwaverandroidapp/TransportDeviceConnectionState.java
/*
 * EMWaver
 * Copyright (c) 2026 Luís Marnoto
 * SPDX-License-Identifier: Apache-2.0
 */

package com.emwaver.emwaverandroidapp;

import androidx.annotation.Nullable;

import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

final class TransportDeviceConnectionState<T> {
    private final T noneTransport;
    private final Map<String, TransportDeviceConnection> connectionsByDeviceId = new HashMap<>();
    private ActiveDeviceTarget<T> target;
    private TransportDeviceConnection connection;

    TransportDeviceConnectionState(T noneTransport) {
        this.noneTransport = noneTransport;
        this.target = new ActiveDeviceTarget<>("active", noneTransport);
    }

    ActiveDeviceTarget<T> setTarget(String deviceId, T transport) {
        target = new ActiveDeviceTarget<>(deviceId, transport);
        connection = null;
        return target;
    }

    void setConnection(@Nullable TransportDeviceConnection connection) {
        this.connection = connection;
        if (connection != null) {
            connectionsByDeviceId.put(normalize(connection.sessionId()), connection);
        }
    }

    void clear() {
        target = new ActiveDeviceTarget<>("active", noneTransport);
        connection = null;
        connectionsByDeviceId.clear();
    }

    void clear(T transport) {
        if (matchesTransport(transport)) {
            clear();
        }
    }

    String currentScriptDeviceId() {
        return connection != null ? connection.sessionId() : target.deviceId;
    }

    boolean matchesDeviceId(String deviceId) {
        return target.matchesDeviceId(deviceId);
    }

    boolean matchesTransport(T transport) {
        return target.matchesTransport(transport);
    }

    T transport() {
        return target.transport;
    }

    @Nullable
    TransportDeviceConnection connection() {
        return connection;
    }

    @Nullable
    TransportDeviceConnection connection(String deviceId) {
        return connectionsByDeviceId.get(normalize(deviceId));
    }

    private static String normalize(String deviceId) {
        String key = deviceId == null || deviceId.trim().isEmpty() ? "active" : deviceId.trim();
        return key.toLowerCase(Locale.US);
    }
}

File: ios/EMWaver/Managers/TransportDeviceConnectionState.swift
/*
 * EMWaver
 * Copyright (c) 2026 Luís Marnoto
 * SPDX-License-Identifier: Apache-2.0
 */

import Foundation

final class TransportDeviceConnectionState<Transport: Equatable> {
    private let noneTransport: Transport
    private var connectionsByDeviceId: [String: TransportDeviceConnection] = [:]
    private var target: ActiveDeviceTarget<Transport>
    private(set) var connection: TransportDeviceConnection?

    init(noneTransport: Transport) {
        self.noneTransport = noneTransport
        self.target = ActiveDeviceTarget(deviceId: "active", transport: noneTransport)
    }

    @discardableResult
    func setTarget(deviceId: String, transport: Transport) -> ActiveDeviceTarget<Transport> {
        let target = ActiveDeviceTarget(deviceId: deviceId, transport: transport)
        self.target = target
        connection = nil
        return target
    }

    func setConnection(_ connection: TransportDeviceConnection?) {
        self.connection = connection
        if let connection {
            connectionsByDeviceId[Self.normalize(connection.sessionKey)] = connection
        }
    }

    func clear() {
        target = ActiveDeviceTarget(deviceId: "active", transport: noneTransport)
        connection = nil
        connectionsByDeviceId.removeAll()
    }

    func clear(transport: Transport) {
        guard matchesTransport(transport) else { return }
        clear()
    }

    var currentScriptDeviceId: String {
        connection?.sessionKey ?? target.deviceId
    }

    var transport: Transport {
        target.transport
    }

    func matchesDeviceId(_ deviceId: String?) -> Bool {
        target.matchesDeviceId(deviceId)
    }

    func matchesTransport(_ transport: Transport) -> Bool {
        target.matchesTransport(transport)
    }

    func connection(for deviceId: String?) -> TransportDeviceConnection? {
        connectionsByDeviceId[Self.normalize(deviceId)]
    }

    private static func normalize(_ deviceId: String?) -> String {
        let key = deviceId?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let normalized = key.isEmpty ? "active" : key
        return normalized.lowercased()
    }
}
